proposals/0000-asset-primitive.md

```diff
@@ -0,0 +1,233 @@
+  <body>
+    <p>Hello, <strong>world!</strong></p>
+    <Icon use={iconCookie} />
+    <Image use={imageBook} />
+  </body>
+</html>
+```
+
+## High-Level Example (Future)
+
+**NOTE: This RFC is not proposing this API!** This is just an example to highlight one kind of high-level API that could be built on top of this primitive in the future.
+
+```astro
+---
+import {Image, Icon, Font} from 'astro/components';
+
+/* Possible Astro.asset() Implementation (for illustration only)
+Astro.asset = function(name: string) {
+  const allAssets = import.meta.globEager('/assets/**');
+  const path = getPathFromName(name);
+  return allAssets[path];
+} */
+---
+<html>
+  <head>
+    <Font use={Astro.asset('roboto-slab-v22-latin-regular')} family="Roboto Slab" />
+    <style>
+      body { font-family: 'Roboto Slab'; }
+    </style>
+  </head>
+  <body>
+    <p>Hello, <strong>world!</strong></p>
+    <Icon use={Astro.asset('book')} />
+    <Image use={Astro.asset('cookie')} />
+  </body>
+</html>
+```
+
+
+
+# Motivation
+
+We have no official, consistent story around loading assets in Astro. Some plugins have attempted to solve this in userland, for example:
+- Icons: https://github.com/natemoo-re/astro-icon
+- Images: https://github.com/Princesseuh/astro-eleventy-img
+- Fonts: https://github.com/aFuzzyBear/astro-ui/tree/main/components/Fonts
+- Fonts: https://gist.github.com/FredKSchott/3d656427d53a3609ae337c60cc6e3002
+
+All of these approaches work today (Astro v0.21), but suffer from an important problem: they resolve assets at **run-time** (aka "during the render"). For example, a component like `<Icon name="book">` can't know the value for `props.name` until the component and page itself is being run and rendered. This is even better illustrated if the value is a variable, like `<Icon name={fetchedAPIUserData.avatarIcon}>`. Astro would never know what this is.
+
+Why is this is a problem? To be clear, it's not a problem *today* because run-time and build-time are the same. In Astro today, **run-time** and **build-time** both happen during your build, ahead-of-time, before the site is deployed to the server. Runtime has full access to your machine, and every source file in your repo.
+
+However, with SSR support coming this will no longer always be true. If `<Icon name={fetchedAPIUserData.avatarIcon}>` renders for the first time on the server (ex: on Netlify, on Vercel) then Astro has no idea which icon you need to include in your build at build-time, which is when the decision of "what to include in your build" is made.
+
+This RFC is a priority to support our community members who are building components and integrations that use assets, and who need an asset strategy that we can commit to working in the future.  There have been different ideas thrown around to solve for this, some of which I will touch on in the "Alternatives" section of this RFC.
+
+### Goals
+
+This proposal outlines an **asset primitive** that accomplishes the following goals:
```

Comment:
What is the specific thing that you would call a primitive? I see two things in this proposal:

- Changing how different asset types are resolved through imports.
- High-level components.

The first is a primitive but the second isn't, imo. If the goal is to outline a primitive maybe the components aren't needed here.

Reply:
Thanks @matthewp, I agree. Removing the components for now to keep scope focused